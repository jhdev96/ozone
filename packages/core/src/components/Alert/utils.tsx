import React from 'react';
import LoadingDots from '../LoadingDots';
import { Info } from '@jhdev96/ozone-icons';
import { Warning } from '@jhdev96/ozone-icons';
import { Success } from '@jhdev96/ozone-icons';
import { Error } from '@jhdev96/ozone-icons';
import Colors from '@jhdev96/ozone-colors';


export type AlertVariation =  "progress"  | 
                              "info"      | 
                              "warning"   | 
                              "success"   | 
                              "error";

export const IconColor = {
  success: Colors.extended.GREEN2,
  error: Colors.extended.RED2,
  info: Colors.extended.BLUE5,
  warning: Colors.extended.YELLOW2,
  progress: Colors.grayScale.BLACK2
}

export const alertIcon = (
  Type: AlertVariation, 
  theme?: 'light' | 'dark'
): JSX.Element => {
  /**
   * Given an alertType and a color theme, 
   * return the appropriate icon for the alert.
   */
  theme ||= 'light';

  switch (Type) {
    case 'success': 
      return (
        <Success 
          color={IconColor.success} 
          className="successIcon"
        />
      );
    case 'warning': 
      return (
        <Warning 
          color={IconColor.warning} 
          className="warningIcon"
        />
      );
    case 'error': 
      return (
        <Error 
          color={IconColor.error} 
          className="errorIcon"
        />
      );
    case 'progress':
      return (
        <LoadingDots 
          theme={
            theme === 'light' 
              ? 'dark' 
              : 'light'
          } 
          position="left" 
          dotSize={6} 
          className="progressIcon"
        />
      );

    default: 
      return (
        <Info 
          color={IconColor.info} 
          className="infoIcon" 
        />
      );
  }
}