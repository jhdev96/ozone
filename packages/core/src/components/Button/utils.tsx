import { LoadingDotsProps } from '../LoadingDots';


export type ButtonVariation = "primary" | "secondary" | "warning" | "basic";

export const loaderTheme = (
  btnType: ButtonVariation, 
  theme: "light" | "dark", 
  outline: boolean
): LoadingDotsProps["theme"] => {
  /** 
   * Sets the theme for the LoadingDots component
   * which is rendered as a child for a button
   * that is loading. 
   * 
   * @param btnType // the type of button wrapping the loader
   * @param theme // the button's theme
   * @param outline // whether the button has just an outline
   * 
   * This function will return
   * an appropriate theme for the LoadingDots
   * component, so that it contrasts with the
   * button.
   */
  switch(btnType) {
    case 'basic':
      if (theme === 'light') return 'dark';
      return 'light';
    case 'primary':
      if (outline) return "blue";
      return theme;
    case 'secondary':
      if (outline) return "red";
      return theme;
    case 'warning':
      return "dark";
    default:
      return theme;
  }
}