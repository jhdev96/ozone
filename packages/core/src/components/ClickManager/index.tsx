import React, { useEffect } from 'react';


interface ClickManagerProps {
  /** 
   * Sets the `className` attribute
   * for the rendered div element.
   */
  className?: string;
  /** 
   * A function that gets called when
   * the rendered element is clicked.
  */
  onClick?: () => void;
  /**
   * A selector used in testing environments.
   */
  dataTestId?: string;
  /** 
   * The component around which this
   * ClickManager will wrap.
   */ 
  children: React.ReactNode;
}

const ClickManager = React.forwardRef<any, ClickManagerProps>(
  function ClickManager({
    className, 
    onClick,
    dataTestId,
    children}: ClickManagerProps, 
    ref: React.ForwardedRef<any>): JSX.Element 
  {
    const localRef = ref as React.MutableRefObject<any>;

    function outsideClick(this: Document, e: MouseEvent) {
      const target = e.target as HTMLElement;
      
      if (localRef && localRef.current && !localRef.current.contains(target)) {
        onClick && onClick();
      }
    }

    useEffect(() => {
      document.addEventListener('click', outsideClick);
      document.addEventListener('contextmenu', outsideClick);
      return () => {
        document.removeEventListener('click', outsideClick);
        document.removeEventListener('contextmenu', outsideClick);
      }
    });

    return (
      <div 
        ref={ref} 
        className={className} 
        data-testid={dataTestId}
      >
        {children}
      </div>
    )
  }
);

ClickManager.displayName = 'ClickManager';

export default ClickManager;