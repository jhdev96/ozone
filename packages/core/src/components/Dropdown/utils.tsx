import React from 'react';
import classNames from '../../utils/classNames';
import styles from './index.module.scss';


export const menuOptions = (
  selectedOption: string, 
  options: (string | React.ReactNode)[], 
  onClick: (e: React.MouseEvent<HTMLDivElement>) => void
): React.ReactNode => {
  /**
    * Return an array of div elements
    * based on the options given by 
    * the client. Each div should 
    * have an onClick prop specifying
    * the function to be invoked when
    * the div or its child element is
    * clicked.
    */
  const optionDivs = options.map((
    option: string | React.ReactNode, 
    idx: React.Key
  ) => (
    <div 
      onClick={onClick} 
      className={styles["menu__option"]} 
      key={idx}
    >
      <span
        className={classNames(
          option === selectedOption && styles.selected
        )}
      >
        {option}
      </span>
    </div>
  ));

  return optionDivs;
};