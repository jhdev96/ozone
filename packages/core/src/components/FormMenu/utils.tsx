import React from 'react';
import styles from './index.module.scss';


export const createTransition = (
  slider: React.MutableRefObject<HTMLDivElement | undefined>, 
  delta: number
) => {
  /** Add a transition to the menu slider when moving to an option
   * that is more than 50 pixels away from the current option.
   * 
   * @param slider // a reference to the menu slider element
   * @param delta // the difference between the two options
   */
  if (Math.abs(delta) > 50) {
    slider.current!.classList.add(styles.transition);

    setTimeout(() => {
      slider.current!.classList.remove(styles.transition)
    }, 400);
  }
}