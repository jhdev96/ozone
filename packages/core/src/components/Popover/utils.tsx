import React from 'react';


const ARROW_SIZE = 10;
export type Position = "top" | "bottom" | "left" | "right";

export const positionPopper = (
  targetRef: React.MutableRefObject<HTMLElement | undefined>, 
  popperRef: React.MutableRefObject<HTMLElement | undefined>, 
  position?: Position
) =>  {
  /**
   * This function helps to position the popover
   * component relative to its target element.
   * @param targetRef // a ref pointing to the target element.
   * @param popperRef // a ref pointing to the popover.
   * @param position // where the popover should be place.
   */
  const target = targetRef.current;
  const popper = popperRef.current;

  switch (position) {
    case 'top':
      if (popper && target) {
        popper!.style.bottom = `${target!.offsetHeight + ARROW_SIZE}px`;
      }
      break;
    case 'left':
    case 'right':
      break;
    default:
      // bottom position
      if (popper && target) {
        popper!.style.top = `${target!.offsetHeight + ARROW_SIZE}px`;
      }
  }
}