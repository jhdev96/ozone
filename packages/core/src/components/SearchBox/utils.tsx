import React from 'react';
import styles from './index.module.scss';


export const toggleFocus = (e: React.FocusEvent<HTMLDivElement>) => {
  const classList = e.currentTarget.classList;
  
  if (classList.contains(styles.focus)) {
    classList.remove(styles.focus);
  } else {
      classList.add(styles.focus);
  }
}