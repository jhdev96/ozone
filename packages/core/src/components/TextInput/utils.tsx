import React from 'react';
import styles from './index.module.scss';


export type InputType = 'email' | 'password' | 'text' | 'number';
export const toggleFocus = (e: React.FocusEvent<HTMLDivElement>) => {
  const classList = e.currentTarget.classList;
  
  if (classList.contains(styles.focus)) {
    classList.remove(styles.focus);
  } else {
      classList.add(styles.focus);
  }
}

export const makeRegexPattern = (Type: InputType): RegExp => {
  /** 
   * Returns a regex pattern against which to validate
   * an input's value.
   */
  switch (Type) {
    case 'email':
      return /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;
    case 'text':
    case 'password':
      return /./;
    case 'number':
      return /^\d+$/;
  }
}

export const withinMinMaxBoundary = (
  min: number | undefined, 
  max: number | undefined, 
  value: number
): boolean => {
  /**
   * Check if a numeric input value
   * is in the allowed range of input
   * values.
   */
  switch (typeof min) {
    case 'undefined':
      if (max === undefined) {
        return true;
      } else {
          return value <= max;
      }
    case 'number':
      if (max !== undefined) {
        return value >= min && value <= max;
      } else {
          return value >= min;
      }
  }
}