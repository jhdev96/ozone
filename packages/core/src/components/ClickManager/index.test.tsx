import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { DummyTestComponent } from '../../utils/test-utils';
import ClickManager from '.';


describe('<ClickManager />', () => {
  function setUpComponent(
    onClick: () => void, 
    ref: React.RefObject<any>
  ) 
  {
    render(
      <div data-testid="wrapper">
        <ClickManager onClick={onClick} ref={ref}>
          <DummyTestComponent />
        </ClickManager>
      </div>
    );
  }

  it('calls `onClick` when the mouse clicks an outside element', () => {
    const ref = React.createRef<HTMLSpanElement>();
    const onClick = jest.fn();

    setUpComponent(onClick, ref);

    const wrapper = screen.getByTestId('wrapper');
    fireEvent.click(wrapper);
    
    expect(onClick).toBeCalled;
  });

  it('calls `onClick` for a right-click outside the component', () => {
    const ref = React.createRef<HTMLSpanElement>();
    const onClick = jest.fn();

    setUpComponent(onClick, ref);

    const wrapper = screen.getByTestId('wrapper');
    fireEvent.contextMenu(wrapper);
    
    expect(onClick).toBeCalled;
  });

  it('passes `classNames` to the wrapper component', () => {
    const {container} = render(
      <ClickManager className="testCss">
        <DummyTestComponent />
      </ClickManager>
    );

    expect(container).toMatchSnapshot();
  });
});