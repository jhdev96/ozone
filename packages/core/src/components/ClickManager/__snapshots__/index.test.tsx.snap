// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClickManager /> passes \`classNames\` to the wrapper component 1`] = `
<div>
  <div
    class="testCss"
  >
    <span
      data-testid="test"
    >
      Hello
    </span>
  </div>
</div>
`;
